examples/solids/qfunctions/finite-strain-neo-hookean-initial-ad.h

```diff
@@ -0,0 +1,494 @@
+                                    ug[1][2][i],
+                                    ug[2][2][i]}
+                                  };
+    // -- Qdata
+    const CeedScalar wdetJ      =   q_data[0][i];
+    const CeedScalar dXdx[3][3] = {{q_data[1][i],
+                                    q_data[2][i],
+                                    q_data[3][i]},
+                                   {q_data[4][i],
+                                    q_data[5][i],
+                                    q_data[6][i]},
+                                   {q_data[7][i],
+                                    q_data[8][i],
+                                    q_data[9][i]}
+                                  };
+    // *INDENT-ON*
+
+    // Compute grad_u
+    //   dXdx = (dx/dX)^(-1)
+    // Apply dXdx to du = grad_u
+    for (CeedInt j = 0; j < 3; j++)     // Component
+      for (CeedInt k = 0; k < 3; k++) { // Derivative
+        grad_u[j][k][i] = 0;
+        for (CeedInt m = 0; m < 3; m++)
+          grad_u[j][k][i] += dXdx[m][k] * du[j][m];
+      }
+
+    // I3 : 3x3 Identity matrix
+    // Compute The Deformation Gradient : F = I3 + grad_u
+    // *INDENT-OFF*
+    const CeedScalar F[3][3] =  {{grad_u[0][0][i] + 1,
+                                  grad_u[0][1][i],
+                                  grad_u[0][2][i]},
+                                 {grad_u[1][0][i],
+                                  grad_u[1][1][i] + 1,
+                                  grad_u[1][2][i]},
+                                 {grad_u[2][0][i],
+                                  grad_u[2][1][i],
+                                  grad_u[2][2][i] + 1}
+                                };
+
+    // *INDENT-OFF*
+    const CeedScalar tempgradu[3][3] =  {{grad_u[0][0][i],
+                                          grad_u[0][1][i],
+                                          grad_u[0][2][i]},
+                                         {grad_u[1][0][i],
+                                          grad_u[1][1][i],
+                                          grad_u[1][2][i]},
+                                         {grad_u[2][0][i],
+                                          grad_u[2][1][i],
+                                          grad_u[2][2][i]}
+                                        };
+    // *INDENT-ON*
+    // E - Green-Lagrange strain tensor
+    //     E = 1/2 (grad_u + grad_u^T + grad_u^T*grad_u)
+    const CeedInt indj[6] = {0, 1, 2, 1, 0, 0}, indk[6] = {0, 1, 2, 2, 2, 1};
+    CeedScalar E2work[6];
+    for (CeedInt m = 0; m < 6; m++) {
+      E2work[m] = tempgradu[indj[m]][indk[m]] + tempgradu[indk[m]][indj[m]];
+      for (CeedInt n = 0; n < 3; n++)
+        E2work[m] += tempgradu[n][indj[m]]*tempgradu[n][indk[m]];
+    }
+
+    // Compute Swork with Enzyme-AD and get tape
+    CeedScalar Swork_[6], packed_tape[6];
+    S_fwd(Swork_, E2work, lambda, mu, packed_tape);
+    for (int j=0; j<6; j++) tape[j][i] = packed_tape[j];
```

Comment:
Hi, @wsmoses. We normally expect the outer loop `for (CeedInt i=0; i<Q; i++)` up above to be vectorized, and thus we'd like the data here to have `i` as its inner index. Clang doesn't do as good a job as gcc vectorizing this variant, but clang and gcc both vectorize the current configuration variants (which we like better anyway, but this one is "harder" in what we ask Enzyme to do). The 6 here comes from `__enzyme_augmentsize` returning `48 = 6 * sizeof(CeedScalar)`, which makes practical sense. If the compiler does its job, this packing will vanish after inlining and the contents of `S_fwd` will used packed instructions. Is there a better way to do this?

I can imagine we add explicit blocking (say batches of 8 elements) and manually write the inner loop, but I'd love to keep that complexity separate.

Reply:
From a pure performance standpoint, I don't think any extra massaging within the derivative to place the tape values at a location would be any different from the packing you have right now (at least in theory). Of course this doesn't mean that we can't make some nice syntactic sugar, though I'm not sure what that would be.

Regarding the size 8 to make it better for vectorization, can we not just round the result of `__enzyme_augmentsize` to the nearest `8*sizeof(CeedScalar)`?

Reply:
What I meant by 8 was blocking over the quadrature point index `i`. Returning `6 * sizeof(CeedScalar)` is totally fine (better because it's less data), but this pack/unpack needs the result of `__enzyme_augmentsize` as a compile-time constant for the loop and data rearrangement to be eliminated. What I meant by unrolling is to do

```c
for (int i=0; i<Q; i+=8) {
   S_fwd(...);
}
```
and `S_fwd` would have its own fixed length-8 loop. I think this clutters the code and at the end of the day doesn't make optimization better. But I see we can assign `__enzyme_augmentsize` to a `static const` and that the optimizer replaces it with a numeric literal so we can have our own packing without magic numbers. So I think we're set, thanks.

Reply:
Oh, or it's sometimes a compile-time constant. Hmm.

Reply:
To clarify, I tried this on a simple function and it worked, but doesn't work here. We could have a run-time or compile-time size compatibility check and just assign `ElasFSInitialNH_AD_tape_bytes = 6 * sizeof(CeedScalar)` here.
```
/home/jed/src/libCEED/examples/solids/problems/../qfunctions/finite-strain-neo-hookean-initial-ad.h:179:53: error: initializer element is not a compile-time constant
static const size_t ElasFSInitialNH_AD_tape_bytes = __enzyme_augmentsize(computeS,
                                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/jed/src/libCEED/examples/solids/problems/finite-strain-neo-hookean-initial-ad.c:29:16: error: initializer element is not a compile-time constant
  .tape_size = ElasFSInitialNH_AD_tape_bytes,
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Reply:
The result of the augmentsize becomes a compile-time constant when the derivatives are generated (which in its current location in the compiler optimization pipeline, is shortly before vectorization). Because it depends on what the definition of the differentiated function is, it isn't available as a constant before then (and thus cannot be constexpr since c++ templates wouldn't be able to tell its value).

That said if you pass a constant as the tape size, Enzyme will check if the memory allocated is at least what it requires and throw an error if not.

Reply:
Okay. I think I'd rather have a run-time error/warning saying to update this variable name to this size than let it fail inside Enzyme. @LeilaGhaffari Could you update your branch to make a constant for the tape size, then get the value produced by `__enzyme_augmentsize` and:
1. if the const size matches, run
2. if the const size is bigger, print a warning that it can be reduced to the run-time size
3. if the const size is smaller, error saying that it must be increased to the run-time size

I think this will be easy enough to maintain and also avoid surprises as we're working on constitutive models. (The checked number is useful information to know.)